### PR TITLE
Sujithbranch

### PR DESCRIPTION
**Pull request by sujithnair1**
spiffs, time sync, encryption and decryption.

### CODE REVIEW

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 1]
```diff
@@ -0,0 +1,82 @@
+#include <AESLib.h>
+
+#define BAUD 9600
+
+AESLib aesLib;
+String kim="r9Huf1ov92qFJMtG";
+// AES Encryption Key
+byte aesKey[] = {kim[0], kim[1], kim[2], kim[3], kim[4], kim[5], kim[6], kim[7], kim[8], kim[9], kim[10], kim[11], kim[12], kim[13], kim[14], kim[15]};
+// General initialization vector (use your own)
+byte aesIv[16] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:10]: Please dont use all zeros, AA, BB etc may be better

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 2]
```diff
@@ -0,0 +1,82 @@
+#include <AESLib.h>
+
+#define BAUD 9600
+
+AESLib aesLib;
+String kim="r9Huf1ov92qFJMtG";
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:6]: how you arrive this key?

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+// AES Encryption Key
+byte aesKey[] = {kim[0], kim[1], kim[2], kim[3], kim[4], kim[5], kim[6], kim[7], kim[8], kim[9], kim[10], kim[11], kim[12], kim[13], kim[14], kim[15]};
+// General initialization vector (use your own)
+byte aesIv[16] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+
+// Generate IV (once)
+void aesInit() {
+  aesLib.gen_iv(aesIv);
+  aesLib.set_paddingmode((paddingMode)0);
+  // encrypt("AAAAAAAAAA", aes_iv); // workaround for incorrect B64 functionality on first run... initing b64 is not enough
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:16]: is this comment required/ relevant? if not please remove

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:17]: space not required

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+}
+String decrypt(char * msg, byte iv[]) {
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:19]: space required before this

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:20]: ditto

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+void aesInit() {
+  aesLib.gen_iv(aesIv);
+  aesLib.set_paddingmode((paddingMode)0);
+  // encrypt("AAAAAAAAAA", aes_iv); // workaround for incorrect B64 functionality on first run... initing b64 is not enough
+
+}
+String decrypt(char * msg, byte iv[]) {
+ 
+  int msgLen = strlen(msg);
+  char decrypted[msgLen]; // half may be enough
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:22]: is this kind of allocation allowed in arduino?

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  aesLib.decrypt64(msg, msgLen, decrypted, aesKey, sizeof(aesKey), iv);
+  Serial.println(decrypted);
+  return String(decrypted);
+}
+
+void setup() {
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:28]: Are you making this as a new project? this should be a new file with out any init ( library) or needed to be added to main controller code

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  Serial.print("free heap: "); Serial.println(ESP.getFreeHeap());
+  #endif
+}
+
+String plaintext = "12345678;";
+int loopcount = 0;
+
+char cleartext[256];
+char ciphertext[512];
+String encrypted="qjGcA6zUUTFkxB7cf/I8/Ot7RsWAPzXxvpjPCySE7GH2nLiIcqTN6Xx304qTExtWMqzVeKrV/yHOYfif1jIsuRhVtF/+9Hp3jMMM1zi3ezS4eOq3om2kwJ7C1ZVZPgIN";
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:48]: What are these? expecting only the functions required for encrypt and decrypt. remove all test stuffs. if this is an example add that in an example folder. as per task, we need couple of library functions ( for encrypt/ decrypt) and key generation algorithms

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+
+
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:51]: so many blank lines

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+void loop() {
+  loopcount++;
+  sprintf(cleartext, "1662428746289472947205789562742762448924269862985629756359768");
```
Comment (greatinbaby) [inline decryptDefaultKey/decryptDefaultKey.ino:54]: again, we dont need any examples here. the supporting functions. this examples can be moved under a folder called examples

File: spiffandtime/spiffandtime.ino [hunk 12]
```diff
@@ -0,0 +1,182 @@
+#include <ESP8266WiFi.h>
```
Comment (greatinbaby) [inline spiffandtime/spiffandtime.ino:1]: make this also as a library file

**Review by greatinbaby — CHANGES_REQUESTED** [review]

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 2 above)
Comment (AbhinandSuresh) [reply decryptDefaultKey/decryptDefaultKey.ino:6]: kim?

File: decryptDefaultKey/decryptDefaultKey.ino [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+  char decrypted[msgLen]; // half may be enough
+  aesLib.decrypt64(msg, msgLen, decrypted, aesKey, sizeof(aesKey), iv);
+  Serial.println(decrypted);
+  return String(decrypted);
+}
+
+void setup() {
+  Serial.begin(BAUD);
+  while(!Serial);
+  delay(2000);
```
Comment (AbhinandSuresh) [inline decryptDefaultKey/decryptDefaultKey.ino:31]: Why this delay? while(!Serial) will be enough right? 
Please Check and update if the delay is really necessary

File: spiffandtime/spiffandtime.ino [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+unsigned long lastsnap;
+unsigned long currenttime;
+unsigned long nowtime;
+unsigned long epoch;
+unsigned long newepoch;
+unsigned long lastHourlySavedTime = 0;
+unsigned long lastDailySavedTime = 0;
+unsigned long lastMonthlySavedTime = 0;
+bool  spiffsActive = false;
+
```
Comment (AbhinandSuresh) [inline spiffandtime/spiffandtime.ino:26]: please avoid these space

File: spiffandtime/spiffandtime.ino [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+    Serial.print("Unable To Find ");
+    Serial.println(filename);
+    Serial.println();
+  }
+}
+void SaveHourlyData( struct water * tankdata) {
+
+  writeFile(HOURLYFILE, tankdata);
+  lastHourlySavedTime = currenttime;
+
```
Comment (AbhinandSuresh) [inline spiffandtime/spiffandtime.ino:114]: please remove spaces

File: spiffandtime/spiffandtime.ino [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+      f.close();
+    }
+  }
+  else {
+    Serial.print("Unable To Find ");
+    Serial.println(filename);
+    Serial.println();
+  }
+}
+void SaveHourlyData( struct water * tankdata) {
```
Comment (AbhinandSuresh) [inline spiffandtime/spiffandtime.ino:110]: please add a space before this line

File: spiffandtime/spiffandtime.ino [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+
+  writeFile(HOURLYFILE, tankdata);
+  lastHourlySavedTime = currenttime;
+
+
+}
+void SaveDialyData() {
```
Comment (AbhinandSuresh) [inline spiffandtime/spiffandtime.ino:117]: please add space here

File: spiffandtime/spiffandtime.ino [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+  struct water tankdata;
+  float average;
+  readfile(HOURLYFILE, &average);
+  tankdata.savetime = currenttime;
+  tankdata.level = average;
+  writeFile(DAILYFILE, &tankdata);
+  lastDailySavedTime = currenttime;
+  Serial.println("data saved in daily file");
+  SPIFFS.remove(HOURLYFILE);
+
```
Comment (AbhinandSuresh) [inline spiffandtime/spiffandtime.ino:127]: please remove this blank line (space)

File: spiffandtime/spiffandtime.ino [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+  tankdata.savetime = currenttime;
+  tankdata.level = average;
+  writeFile(MONTHLYFILE, &tankdata);
+  lastMonthlySavedTime = currenttime;
+  Serial.println("data saved in daily file");
+  SPIFFS.remove(DAILYFILE);
+}
+
+void setup() {
+
```
Comment (AbhinandSuresh) [inline spiffandtime/spiffandtime.ino:143]: please remove this space

**Review by AbhinandSuresh — CHANGES_REQUESTED** [review]

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 1 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:10]: ok

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 2 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:6]: ok will use meaning variable names and will not use default key

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 3 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:16]: will remove

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 4 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:17]: ok

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 8 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:28]: will create seperate programs for examples and functions

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 9 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:48]: ok

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 11 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:54]: ok

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 10 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:51]: will remove blank lines

File: decryptDefaultKey/decryptDefaultKey.ino (hunk 13 above)
Comment (sujithnair1) [reply decryptDefaultKey/decryptDefaultKey.ino:31]: ok

File: spiffandtime/spiffandtime.ino (hunk 14 above)
Comment (sujithnair1) [reply spiffandtime/spiffandtime.ino:26]: ok

File: spiffandtime/spiffandtime.ino (hunk 16 above)
Comment (sujithnair1) [reply spiffandtime/spiffandtime.ino:110]: ok

File: spiffandtime/spiffandtime.ino (hunk 18 above)
Comment (sujithnair1) [reply spiffandtime/spiffandtime.ino:127]: ok

File: spiffandtime/spiffandtime.ino (hunk 19 above)
Comment (sujithnair1) [reply spiffandtime/spiffandtime.ino:143]: ok

**Pull request closed without merging**